Improve strategy development documentation (#10779)
freqtrade/freqtrade: docs/freqtrade-101.md

# Freqtrade Strategies 101: A Quick Start for Strategy Development

For the purposes of this quick start, we are assuming you are familiar with the basics of trading.

## Required Knowledge

A strategy in Freqtrade is a Python class that defines the logic for buying and selling cryptocurrency `assets`.

> Assets are defined as `pairs`, which represent the `coin` and the `stake`. The coin is the asset you are trading using another currency as the stake.

Reviewer: we might want to reference freqtrade basics here ? it covers this topic in more depth ... ? (unless we want to explicitly keep it concise and "non-referencing" - so readers can stay on this page).

Data is supplied by the exchange in the form of `candles`, which are made up of a six values: `date`, `open`, `high`, `low`, `close` and `volume`.

`Technical analysis` functions analyse the candle data using various computational and statistical formulae, and produce secondary values called `indicators`.

Indicators are analysed on the asset pair candles to generate `signals`.

Signals are turned into `orders` on a cryptocurrency `exchange`, i.e. `trades`.

We use the terms `entry` and `exit` instead of `buying` and `selling` because Freqtrade supports both `long` and `short` trades.

> - long:
>   - You buy the coin based on a stake, e.g. buying the coin BTC using USDT as your stake, and you make a profit by selling the coin at a higher rate than you paid for. Profits are made in long trades by the coin value going up versus the stake.
> - short
>   - You borrow capital from from the exchange in the form of the coin, and you pay back the stake value of the coin later. Profits are made in short trades by the coin value going down versus the stake (you pay the loan off at a lower rate).

Reviewer: I'm not sure i like the nested list here (from a rendering perspective) - it kinda breaks reading flow 
maybe either write long/short bold (instead of as list) - or have 1 level of lists (something like the below?)

```suggestion
- **long**: You buy the coin based on a stake, e.g. buying the coin BTC using USDT as your stake, and you make a profit by selling the coin at a higher rate than you paid for. Profits are made in long trades by the coin value going up versus the stake.
- **short**: You borrow capital from from the exchange in the form of the coin, and you pay back the stake value of the coin later. Profits are made in short trades by the coin value going down versus the stake (you pay the loan off at a lower rate).
```

> For simplicity, here we will focus on spot (long) trades only.

Reviewer: this kinda contradicts the below "hint" boxes?

Author: It was "here are the basic terms, but because shorting is more complicated, we'll focus on longs"

## Structure of a Basic Strategy

### Main dataframe

Freqtrade strategies use a tabular data structure with rows and columns known as a `dataframe` to generate signals to enter and exit trades.

> Freqtrade dataframes are organised by pair, and are indexed by the `date` column, e.g. `2024-06-31 12:00`.

Reviewer: is organised UK english? spell-check doesn't know UK english apparently :D 

![image](https://github.com/user-attachments/assets/d6c6174e-18cc-4235-8c6f-d8368cbe7be1)

Just a random observation - i don't really care "which" english we use, as long as it's either UK or US - and not "finish english" (or whatever other country's english) :laughing:

Author: Sorry, yup, British English. I can change it to American *I suppose* 😄

Reviewer: don't bother learning a new foreign language :laughing:

The next 5 columns represent the `open`, `high`, `low`, `close` and `volume` (OHLCV) data.

### Populate indicator values
The `populate_indicators` function adds columns to the dataframe that represent the technical analysis indicator values.

Examples of common indicators include Relative Strength Index, Bollinger Bands, Money Flow Index, Moving Average, and Average True Range.

> Columns are added to the dataframe by calling technical analysis functions, e.g. ta-lib's RSI function `ta.RSI()`, and assigning them to a column name, e.g. `rsi`

Reviewer: Is it worth hinting at the fact that not all indicators (also depending on the library) return a series - so this approach may need to be adjusted depending on the library / indicator?
Could also be a case of "no, let's keep it simple" ;)

Author: Yeah I'd prefer simplicity for a quick start.

```python
    dataframe['rsi'] = ta.RSI(dataframe)
```

### Populate entry signals
The `populate_entry_trend` function defines conditions for an entry signal.

The dataframe column `enter_long` is added to the dataframe, and when a value of `1` is in this column, Freqtrade sees an entry signal.

??? Hint "Shorting"

>     When shorting, this column is called `enter_short`.

Reviewer: ```suggestion
    To enter short trades, use the column name `enter_short`.
```

?

### Populate exit signals
The `populate_exit_trend` function defines conditions for an exit signal.

The dataframe column `exit_long` is added to the dataframe, and when a value of `1` is in this column, Freqtrade sees an exit signal.

??? Hint "Shorting"
    When shorting, this column is called `exit_short`.

## A simple strategy

Here is a minimal example of a Freqtrade strategy:

```python
from freqtrade.strategy.interface import IStrategy

> from pandas import DataFrame

Reviewer: misses the ta import ... (i think this should keep default imports - for simplified "copy/paste" as starting point).

Author: Good spot!

class MyStrategy(IStrategy):

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # generate values for technical analysis indicators
        dataframe['rsi'] = ta.RSI(dataframe, timeperiod=14)

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # generate entry signals based on indicator values
        dataframe.loc[
            (dataframe['rsi'] < 30),
            'enter_long'] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # generate exit signals based on indicator values
        dataframe.loc[
            (dataframe['rsi'] > 70),
            'exit_long'] = 1

        return dataframe
```

## Making trades

> When a signal is found (a `1` in an entry or exit column), Freqtrade will attempt to make an order on the exchange, i.e. a `trade`.

Reviewer: I'd extend this a bit mentioning that `signal != order` ... as that's far from a given.

Author: OK good point. I'll add that.

[Freqtrade basics](bot-basics.md) page.

## Required Knowledge

A strategy in Freqtrade is a Python class that defines the logic for buying and selling cryptocurrency `assets`.

Assets are defined as `pairs`, which represent the `coin` and the `stake`. The coin is the asset you are trading using another currency as the stake.

Data is supplied by the exchange in the form of `candles`, which are made up of a six values: `date`, `open`, `high`, `low`, `close` and `volume`.

`Technical analysis` functions analyse the candle data using various computational and statistical formulae, and produce secondary values called `indicators`.

Indicators are analysed on the asset pair candles to generate `signals`.

Signals are turned into `orders` on a cryptocurrency `exchange`, i.e. `trades`.

We use the terms `entry` and `exit` instead of `buying` and `selling` because Freqtrade supports both `long` and `short` trades.

- **long**: You buy the coin based on a stake, e.g. buying the coin BTC using USDT as your stake, and you make a profit by selling the coin at a higher rate than you paid for. Profits are made in long trades by the coin value going up versus the stake.

> - **short**: You borrow capital from the exchange in the form of the coin, and you pay back the stake value of the coin later. Profits are made in short trades by the coin value going down versus the stake (you pay the loan off at a lower rate).

Reviewer: `Profits are made in long trades by the coin value going up versus the stake.`
the 2nd sentence each sounds a bit odd ... can't tell exactly why - maybe "in long trades, profits are made..." would sound better?
I'm also not sure if "going up versus the stake" couldn't be rephrased into "by the price of the coin raising" ?

```python
dataframe['rsi'] = ta.RSI(dataframe)
```

??? Hint "Technical Analysis libraries"
    Different libraries work in different ways to generate indicator values. Please check the documentation of each library to understand
    how to integrate it into your strategy. You can also check the [Freqtrade example strategies](https://github.com/freqtrade/freqtrade-strategies) to give you ideas.

### Populate entry signals

The `populate_entry_trend` function defines conditions for an entry signal.

The dataframe column `enter_long` is added to the dataframe, and when a value of `1` is in this column, Freqtrade sees an entry signal.

??? Hint "Shorting"
    To enter short trades, use the `enter_short` column.

### Populate exit signals

The `populate_exit_trend` function defines conditions for an exit signal.

The dataframe column `exit_long` is added to the dataframe, and when a value of `1` is in this column, Freqtrade sees an exit signal.

??? Hint "Shorting"
    To exit short trades, use the `exit_short` column.

## A simple strategy

Here is a minimal example of a Freqtrade strategy:

```python

> from freqtrade.strategy.interface import IStrategy

Reviewer: ```suggestion
from freqtrade.strategy import IStrategy
```

from pandas import DataFrame
import talib.abstract as ta

class MyStrategy(IStrategy):

>     def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

Reviewer: I think at least `stoploss`, and maybe `minimal_roi` should be added. Without stoploss, the strategy won't start unless you're using a non-default config (with stoploss configured there).
This should however be a running strategy in my eyes...

        # generate values for technical analysis indicators
        dataframe['rsi'] = ta.RSI(dataframe, timeperiod=14)

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # generate entry signals based on indicator values
        dataframe.loc[
            (dataframe['rsi'] < 30),
            'enter_long'] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # generate exit signals based on indicator values
        dataframe.loc[
            (dataframe['rsi'] > 70),
            'exit_long'] = 1

        return dataframe
```

## Making trades

When a signal is found (a `1` in an entry or exit column), Freqtrade will attempt to make an order, i.e. a `trade` or `position`.

The number of concurrent trades that can be opened is defined by the `max_open_trades` [configuration](configuration.md) option.

However, there can be a range of scenarios where generating a signal does not always create a trade order. These include:

- not enough remaining stake to buy an asset, or funds in your wallet to sell an asset (including any fees)

> - not enough open slots for a new trade to be opened

Reviewer: (reading with an unknowing mind ...) 
Above we talked about "concurrent trades" - so what's slots again?

- there is already an open trade for a pair (Freqtrade cannot stack positions - however it can [adjust existing positions](strategy-callbacks.md#adjust-trade-position))

> - if an entry and exit signal is present on the same candle, they cancel each other out and no order will be raised

Reviewer: call it (and link) to colliding signals?

We use the terms `entry` and `exit` instead of `buying` and `selling` because Freqtrade supports both `long` and `short` trades.

- **long**: You buy the coin based on a stake, e.g. buying the coin BTC using USDT as your stake, and you make a profit by selling the coin at a higher rate than you paid for. Profits are made in long trades by the coin value going up versus the stake.

> - **short**: You borrow capital from from the exchange in the form of the coin, and you pay back the stake value of the coin later. Profits are made in short trades by the coin value going down versus the stake (you pay the loan off at a lower rate).

Reviewer: ```suggestion
- **short**: You borrow capital from the exchange in the form of the coin, and you pay back the stake value of the coin later. Profits are made in short trades by the coin value going down versus the stake (you pay the loan off at a lower rate).
```

Dry runs are enabled by setting `dry_run` to true in your [configuration](configuration.md#using-dry-run-mode).

!!! Warning "Backtests can be very inaccurate"
    There are many reasons why backtest results will not match reality. Please check the [backtesting assumptions](backtesting.md#assumptions-made-by-backtesting) and [common strategy mistakes](strategy-customization.md#common-mistakes-when-developing-strategies) documentation.

!!! Warning "Public strategies can have significant issues"
    There are many websites listing strategies with impressive backtest results. Do not assume these results are achieveable or realistic.

??? Hint "Useful commands"
    Freqtrade includes two useful commands to check for basic flaws in strategies: [lookahead-analysis](lookahead-analysis.md) and [recursive-analysis](recursive-analysis.md).

> !!! Note "Always dry run first!"

Reviewer: I found that putting several boxes in sequence reduces the likelyhood of readers actually reading the content with every additional box.

maybe we can combine the 2 warnings into one bigger box?

Author: Done

>     Always dry run your strategy after backtesting it to see if backtesting and dry run results match, giving you confidence that things are operating correctly.

Reviewer: i think this suggests (a bit too much) that "backtest and dry-run results must match" - but you'll very rarely find 2 100% matching sets ... 
Maybe slightly different wording suggesting "mostly match" or similar might work better? (will let an english native determine exact wording though)

Reviewer 2: "dry run and backtest results are similar. If there is any difference, you need to do some investigation and make sure that the difference isn't the enter and exit signals. It's normal if ROI, stoploss, trailing stoploss and custom exit/stoploss differ due to different entry rate between dry run and backtest".

And then either explain why the entry rate can differ in the next paragraph or in a separate section
